File: ts/test/entity/v2018/V2018Direct.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import { test, describe } from 'node:test'
import assert from 'node:assert'


import { AareguruSDK } from '../../..'

import {
  envOverride,
} from '../../utility'


describe('V2018Direct', async () => {

  test('direct-exists', async () => {
    const sdk = new AareguruSDK({
      system: { fetch: async () => ({}) }
    })
    assert('function' === typeof sdk.direct)
    assert('function' === typeof sdk.prepare)
  })


  test('direct-load-v2018', async () => {
    const setup = directSetup({ id: 'direct01' })
    const { client, calls } = setup

    const params: any = {}
    if (!setup.live) {

    }

    const result: any = await client.direct({
      path: 'v2018/history',
      method: 'GET',
      params,
    })

    assert(result.ok === true)
    assert(result.status === 200)
    assert(null != result.data)

    if (!setup.live) {
      assert(result.data.id === 'direct01')
      assert(calls.length === 1)
      assert(calls[0].init.method === 'GET')
    }
  })

})



function directSetup(mockres?: any) {
  const calls: any[] = []

  const env = envOverride({
    'AAREGURU_TEST_V_____ENTID': {},
    'AAREGURU_TEST_LIVE': 'FALSE',
    'AAREGURU_APIKEY': 'NONE',
  })

  const live = 'TRUE' === env.AAREGURU_TEST_LIVE

  if (live) {
    const client = new AareguruSDK({
      apikey: env.AAREGURU_APIKEY,
    })

    let idmap: any = env['AAREGURU_TEST_V_____ENTID']
    if ('string' === typeof idmap && idmap.startsWith('{')) {
      idmap = JSON.parse(idmap)
    }

    return { client, calls, live, idmap }
  }

  const mockFetch = async (url: string, init: any) => {
    calls.push({ url, init })
    return {
      status: 200,
      statusText: 'OK',
      headers: {},
      json: async () => (null != mockres ? mockres : { id: 'direct01' }),
    }
  }

  const client = new AareguruSDK({
    base: 'http://localhost:8080',
    system: { fetch: mockFetch },
  })

  return { client, calls, live, idmap: {} as any }
}
  

File: lua/test/v2018_direct_test.lua
-- V2018 direct test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("aareguru_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

describe("V2018Direct", function()
  it("should direct-load-v2018", function()
    local setup = v2018_direct_setup({ id = "direct01" })
    local client = setup.client


    local result, err = client:direct({
      path = "v2018/history",
      method = "GET",
      params = {},
    })
    assert.is_nil(err)
    assert.is_true(result["ok"])
    assert.are.equal(200, helpers.to_int(result["status"]))
    assert.is_not_nil(result["data"])

    if not setup.live then
      if type(result["data"]) == "table" then
        assert.are.equal("direct01", result["data"]["id"])
      end
      assert.are.equal(1, #setup.calls)
    end
  end)

end)


function v2018_direct_setup(mockres)
  runner.load_env_local()

  local calls = {}

  local env = runner.env_override({
    ["AAREGURU_TEST_V_____ENTID"] = {},
    ["AAREGURU_TEST_LIVE"] = "FALSE",
    ["AAREGURU_APIKEY"] = "NONE",
  })

  local live = env["AAREGURU_TEST_LIVE"] == "TRUE"

  if live then
    local merged_opts = {
      apikey = env["AAREGURU_APIKEY"],
    }
    local client = sdk.new(merged_opts)
    return {
      client = client,
      calls = calls,
      live = true,
      idmap = {},
    }
  end

  local function mock_fetch(url, init)
    table.insert(calls, { url = url, init = init })
    return {
      status = 200,
      statusText = "OK",
      headers = {},
      json = function()
        if mockres ~= nil then
          return mockres
        end
        return { id = "direct01" }
      end,
      body = "mock",
    }, nil
  end

  local client = sdk.new({
    base = "http://localhost:8080",
    system = {
      fetch = mock_fetch,
    },
  })

  return {
    client = client,
    calls = calls,
    live = false,
    idmap = {},
  }
end
